bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/EclipsePreferences.java

```diff
@@ -629,15 +628,26 @@ protected String internalPut(String key, String newValue) {
 			// illegal state if this node has been removed
 			checkRemoved();
 			String oldValue = properties.get(key);
-			if (oldValue != null && oldValue.equals(newValue))
+			if (oldValue != null && oldValue.equals(newValue)) {
 				return oldValue;
-			if (DEBUG_PREFERENCE_SET)
+			} else if (DEBUG_PREFERENCE_SET) {
 				PrefsMessages.message("Setting preference: " + absolutePath() + '/' + key + '=' + newValue); //$NON-NLS-1$
-			properties = properties.put(key, newValue);
+			}
+			properties = properties.put(intern(key), intern(newValue));
 			return oldValue;
 		}
 	}
 
+	private static final Map<String, String> STRING_POOL = new WeakHashMap<>();
```

Comment:
I believe this will pin the String for all time because `WeakHashMap` only keeps weak references to the keys.  The values continue to have strong references if I recall correctly.  So using the cached string as both the key and the value will causing that string to be pinned and never removed.  I'm unsure why we don't just use `String.intern`.  I'm fairly certain modern VMs have solved the old performance concerns with using intern.

Reply:
While agreeing with the first part, I'm not confident that "modern VMs have solved the old performance concerns with using intern." I only trust facts & numbers :-)

Regarding the first part, see also yet another place where we do String caching: `org.eclipse.jdt.internal.core.util.DeduplicationUtil`